### PR TITLE
fix(sync): fix testnet syncer loop on large Orchard blocks

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Testnet block 1860741 contains a 600 kB Orchard transaction, which can take a long time to verify.

This causes a syncer loop:
- the syncer requests that block again
- which causes a duplicate request error,
- which restarts the syncer loop,
- which re-verifies the same block.

Eventually, ~~all the CPUs are busy verifying the same block,~~ and Zebra does not make any sync progress.

Edit: actually I think there's just one verification going at one time. The CPU bug could be something different.

## Solution

- Ignore the duplicate request error, and allow the block to verify
    - Make all download errors into enum variants
    - Check all syncer request errors to see if we want to ignore them

Related fixes:
- Check genesis errors to log info or warn messages

We might also need to increase the verification timeout on slower machines (PR #4156).

## Review

This bug blocks syncing on testnet past block 1860741, so it is a high priority.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Better syncer errors

### CODE REVIEW

File: zebrad/src/components/sync.rs [hunk 1]
```diff
@@ -860,6 +884,14 @@ where
...
                 false
             }
+            BlockDownloadVerifyError::DuplicateBlockQueuedForDownload { .. } => {
+                debug!(
+                    error = ?e,
+                    "queued duplicate block hash for download, \
+                     assuming the syncer will eventually resolve duplicates, continuing"
+                );
+                false
+            }
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:905]: This ignores the duplicate queued block hash error

File: zebrad/src/components/sync.rs [hunk 2]
```diff
@@ -887,13 +921,12 @@ where
...
                 // to write a test to check it. The test below is a best-effort
                 // attempt to catch if that happens and log it.
+                //
                 // TODO: add a proper test and remove this
                 // https://github.com/ZcashFoundation/zebra/issues/2909
                 let err_str = format!("{:?}", e);
                 if err_str.contains("AlreadyVerified")
                     || err_str.contains("AlreadyInChain")
-                    || err_str.contains("Cancelled")
-                    || err_str.contains("BehindTipHeight")
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:896]: These are structural matches, so we don't need to check strings as well

File: zebrad/src/components/sync.rs [hunk 3]
```diff
@@ -748,14 +750,28 @@ where
...
+                        warn!(
+                            ?error,
+                            "could not download or verify genesis block, retrying"
+                        );
+                    } else {
+                        info!(
+                            ?error,
+                            "temporary error downloading or verifying genesis block, retrying"
+                        );
+                    }
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:773]: This lowers the log severity for temporary genesis errors

File: zebrad/src/components/sync.rs [hunk 4]
```diff
@@ -594,7 +594,8 @@ where
         // so the last peer to respond can't toggle our mempool
         self.recent_syncs.push_obtain_tips_length(new_downloads);
 
-        self.request_blocks(download_set).await?;
+        let response = self.request_blocks(download_set).await;
+        Self::handle_response(response)?;
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:598]: This ignores temporary obtaintips errors

File: zebrad/src/components/sync.rs [hunk 5]
```diff
@@ -731,7 +732,8 @@ where
         // so the last peer to respond can't toggle our mempool
         self.recent_syncs.push_extend_tips_length(new_downloads);
 
-        self.request_blocks(download_set).await?;
+        let response = self.request_blocks(download_set).await;
+        Self::handle_response(response)?;
```
Comment (teor2345) [inline zebrad/src/components/sync.rs:736]: This ignores temporary extendtips errors

**Review by oxarbitrage — APPROVED** [review]
Looks good, should help.

File: zebrad/src/components/sync.rs (hunk 5 above)
Comment (oxarbitrage) [reply zebrad/src/components/sync.rs:736]: Thanks for the explanation comments.

**Review by dconnolly — COMMENTED** [review]
👍

**Pull request merged**